### PR TITLE
Add player movement properties

### PR DESCRIPTION
**Pull request by Gamersxp6248**
:setRunSpeed(number)
:setWalkSpeed(number)
etc

### CODE REVIEW

**Comment by thegrb93** [comment]
I imagine most of the setters need to be clamped and nan checked.

**Comment by Gamersxp6248** [comment]
Surprisingly only about two or three of them needed it as far as I checked, I'll make sure to double check for you.

**Comment by Gamersxp6248** [comment]
Using 0 is basically the same as every negative value, just gives you immense speed. so not really anything to worry about

File: lua/starfall/libs_sh/players.lua [hunk 1]
```diff
@@ -618,6 +668,116 @@ if SERVER then
...
+	end
+	
+	--- Sets Duck Speed
+	-- @server
+	-- @param number newduckspeed New Duck speed, This is a multiplier from 0 to 1.
+	function player_methods:setDuckSpeed(val)
+		local ent = getply(self)
+		checkpermission(instance, ent, "player.modifyMovementProperties")
+		checkluatype(val, TYPE_NUMBER)
+		ent:SetDuckSpeed(math.Clamp(val,0.005,0.995))
```
Comment (thegrb93) [inline lua/starfall/libs_sh/players.lua:689]: Wouldn't a value greater than 1 also be wanted in some cases? I'd say 0.001 to 1000 would be acceptable clamping range.

File: lua/starfall/libs_sh/players.lua [hunk 2]
```diff
@@ -618,6 +668,116 @@ if SERVER then
...
+	end
+	
+	--- Sets Ladder Climb Speed, probably unstable
+	-- @server
+	-- @param number newladderclimbspeed New Ladder Climb speed.
+	function player_methods:setLadderClimbSpeed(val)
+		local ent = getply(self)
+		checkpermission(instance, ent, "player.modifyMovementProperties")
+		checkluatype(val, TYPE_NUMBER)
+		ent:SetLadderClimbSpeed(math.max(val,0))
```
Comment (thegrb93) [inline lua/starfall/libs_sh/players.lua:709]: Needs an upper bound. Maybe 0 to 10000?

File: lua/starfall/libs_sh/players.lua [hunk 3]
```diff
@@ -618,6 +668,116 @@ if SERVER then
...
+	end
+	
+	--- Sets Max Speed
+	-- @server
+	-- @param number newmaxspeed New Max speed.
+	function player_methods:setMaxSpeed(val)
+		local ent = getply(self)
+		checkpermission(instance, ent, "player.modifyMovementProperties")
+		checkluatype(val, TYPE_NUMBER)
+		ent:SetMaxSpeed(math.max(val,0))
```
Comment (thegrb93) [inline lua/starfall/libs_sh/players.lua:719]: Upper bound needed

Comment (thegrb93) [reply lua/starfall/libs_sh/players.lua:719]: I also need to check if clamp handles nan as well.

File: lua/starfall/libs_sh/players.lua (hunk 1 above)
Comment (Gamersxp6248) [reply lua/starfall/libs_sh/players.lua:689]: https://github.com/Facepunch/garrysmod-issues/issues/2722

**Review by thegrb93 — CHANGES_REQUESTED** [review]
Looks like clamp unfortunately doesn't stop nans from getting through. Can you do nan checks too? Should be as easy as:
`math.max(val==val and val or 0, 0)`

**Comment by Gamersxp6248** [comment]
> Looks like clamp unfortunately doesn't stop nans from getting through. Can you do nan checks too? Should be as easy as: `val==val and math.max(val, 0) or 0`




checkluatype(val, TYPE_NUMBER)

**Comment by thegrb93** [comment]
> I hope I haven't misunderstood and took "Not A Number" literally, and there's some other vague thing you are referring to such as infinite exponents or something similar.

Yes, you did unfortunately lol. nan is what you get when you 0/0 or sqrt(-1).
I was thinking about adding a 'checkvalidnumber' function which would do this check.

**Comment by thegrb93** [comment]
If you can replace the checkluatype with SF.CheckValidNumber then I think it should work. Make sure you update your branch with latest master.

**Pull request merged**